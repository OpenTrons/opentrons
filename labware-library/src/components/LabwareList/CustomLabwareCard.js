// @flow
// labware creator display card
import * as React from 'react'
import { Link } from 'react-router-dom'
import { getPublicPath } from '../../public-path'
import styles from './styles.css'

type Props = {
  isResultsEmpty?: boolean,
}
const SUPPORT_URL =
  'https://support.opentrons.com/en/articles/3136504-creating-custom-labware-definitions'

export function CustomLabwareCard(props: Props): React.Node {
  const PROMPT_TEXT = props.isResultsEmpty
    ? 'No results found'
    : "Don't see your labware here?"
  return (
    <li className={styles.custom_labware_card}>
      <span className={styles.custom_labware_text}>{PROMPT_TEXT}</span>

      <a
        className={styles.btn_white}
        href={SUPPORT_URL}
        target="_blank"
        rel="noopener noreferrer"
      >
        Learn more about creating custom definitions
      </a>
      <Link to={`${getPublicPath()}create`} className={styles.btn_blue}>
        Go to the labware creator
      </Link>
    </li>
  )
}
