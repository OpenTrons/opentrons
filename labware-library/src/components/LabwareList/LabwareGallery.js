// @flow
import * as React from 'react'

import { LabwareRender, RobotWorkSpace } from '@opentrons/components'
import styles from './styles.css'

import type { LabwareDefinition } from '../../types'

export type GalleryProps = {
  definition: LabwareDefinition,
}

export type GalleryState = {
  currentImage: number,
}

type ThumbnailProps = {
  onClick: () => mixed,
  children: React.Node,
}

class Gallery extends React.Component<GalleryProps, GalleryState> {
  constructor(props: GalleryProps) {
    super(props)
    this.state = { currentImage: 1 }
  }

  render() {
    const { definition } = this.props
    const { currentImage } = this.state

    // TODO(mc, 2019-03-27): use actual images
    const images = [
      <img key="left" src={`https://placekitten.com/480/480`} />,
      <RobotWorkSpace
        key="center"
        viewBox={`0 0 ${definition.dimensions.xDimension} ${
          definition.dimensions.yDimension
        }`}
      >
        {() => <LabwareRender definition={definition} />}
      </RobotWorkSpace>,
      <img key="right" src={`https://placekitten.com/512/512`} />,
    ]

    return (
      <div className={styles.gallery}>
        <div className={styles.gallery_main}>
          <div className={styles.image_container}>{images[currentImage]}</div>
        </div>
        <div className={styles.thumbnail_row}>
          {images.map((img, index) => (
            <Thumbnail
              key={index}
              imageId={index}
              onClick={() => this.setState({ currentImage: index })}
            >
              {img}
            </Thumbnail>
          ))}
        </div>
      </div>
    )
  }
}

function Thumbnail(props: ThumbnailProps) {
  const { onClick, children } = props

  return (
    <button className={styles.thumbnail_container} onClick={onClick}>
      <div className={styles.thumbnail}>
        <div className={styles.image_container}>{children}</div>
      </div>
    </button>
  )
}

export default Gallery
