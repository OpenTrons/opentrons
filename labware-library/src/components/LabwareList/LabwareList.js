// @flow
// main LabwareList component
import * as React from 'react'

import styles from './styles.css'

import {getAllDefinitions} from '../../definitions'
import LabwareCard from './LabwareCard'

export default function LabwareList () {
  return (
    <ul className={styles.list}>
      {getAllDefinitions().map(d => (
        <LabwareCard key={d.otId} definition={d} />
      ))}
    </ul>
  )
}
