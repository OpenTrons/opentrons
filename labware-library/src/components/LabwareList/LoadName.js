// @flow
// labware load name with copy button
import * as React from 'react'

import { IconButton, Tooltip } from '@opentrons/components'
import styles from './styles.css'

// TODO(mc, 2019-03-29): i18n
const EN_API_NAME = 'api name'
const EN_COPY_SUCCESS_MESSAGE = 'Copied to clipboard!'

const COPY_ICON = 'ot-copy-text'
const SUCCESS_TIMEOUT_MS = 1500

export type LoadNameProps = {
  loadName: string,
}

export type LoadNameState = {
  success: boolean,
}

class LoadName extends React.Component<LoadNameProps, LoadNameState> {
  inputRef: { current: HTMLInputElement | null }
  successTimeout: TimeoutID | null

  constructor(props: LoadNameProps) {
    super(props)
    this.inputRef = React.createRef()
    this.successTimeout = null
    this.state = { success: false }
  }

  // note: we could choose to always copy the entire loadName string here,
  // regardless of what the user selects, but the benefit of catching missed
  // characters doesn't seem to outweigh the annoyance of removing user control
  handleCopy = () => {
    this.setState({ success: true })
    this.cleanupSuccessTimeout()
    this.successTimeout = setTimeout(
      () => this.setState({ success: false }),
      SUCCESS_TIMEOUT_MS
    )
  }

  handleCopyButtonClick = () => {
    if (this.inputRef.current) {
      this.inputRef.current.select()
      document.execCommand('copy')
    }
  }

  cleanupSuccessTimeout() {
    if (this.successTimeout) clearTimeout(this.successTimeout)
  }

  componentWillUnmount() {
    this.cleanupSuccessTimeout()
  }

  render() {
    const { loadName } = this.props
    const { success } = this.state

    return (
      <div className={styles.load_name}>
        <label className={styles.load_name_label} onCopy={this.handleCopy}>
          <p className={styles.top_label}>{EN_API_NAME}</p>
          <input
            ref={this.inputRef}
            className={styles.load_name_input}
            type="text"
            value={loadName}
            onFocus={e => e.currentTarget.select()}
            readOnly
          />
        </label>
        <Tooltip open={success} tooltipComponent={EN_COPY_SUCCESS_MESSAGE}>
          {tooltipProps => (
            <IconButton
              onClick={this.handleCopyButtonClick}
              hoverTooltipHandlers={tooltipProps}
              className={styles.load_name_button}
              name={COPY_ICON}
              inverted
            />
          )}
        </Tooltip>
      </div>
    )
  }
}

export default LoadName
