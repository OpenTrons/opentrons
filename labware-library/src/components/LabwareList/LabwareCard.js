// @flow
// labware display card
// TODO(mc, 2019-03-21): split this file out into multiple component files,
//   many of which will be common to LabwareCard and LabwarePage
import * as React from 'react'
import { Link } from 'react-router-dom'
import round from 'lodash/round'

import { getDisplayVolume } from '@opentrons/shared-data'
import { getPublicPath } from '../../public-path'
import { getUniqueWellProperties } from '../../definitions'
import { Icon } from '@opentrons/components'
import Gallery from './LabwareGallery'
import LoadName from './LoadName'
import Tags from './Tags'
import styles from './styles.css'

import {
  LabeledValueTable,
  LabelText,
  Value,
  TABLE_ROW,
  LABEL_TOP,
} from '../ui'

import {
  CATEGORY_LABELS_BY_CATEGORY,
  NUM_WELLS_SHORT_BY_CATEGORY,
  LABWARE_DIMS_BY_CATEGORY,
  WELL_DIMS_BY_CATEGORY,
  MM,
  SHORT_X_DIM,
  SHORT_Y_DIM,
  SHORT_Z_DIM,
  X_DIM,
  Y_DIM,
  DIAMETER,
  DEPTH,
  MAX_VOLUME,
} from '../../localization'

import type { LabwareDefinition } from '../../types'

// safe toFixed
const toFixed = (n: number): string => round(n, 2).toFixed(2)

export type LabwareCardProps = { definition: LabwareDefinition }

export default function LabwareCard(props: LabwareCardProps) {
  const { definition } = props

  return (
    <li className={styles.card}>
      <TopBar {...props} />
      <Title {...props} />
      <div className={styles.gallery_container}>
        <Gallery {...props} />
      </div>
      <div className={styles.stats}>
        <Dimensions {...props} />
        <Wells {...props} />
        <WellProperties {...props} />
      </div>
      <div className={styles.tags_container}>
        <Tags {...props} />
      </div>
      <LoadName loadName={definition.parameters.loadName} />
    </li>
  )
}

function TopBar(props: LabwareCardProps) {
  const { metadata, brand } = props.definition

  return (
    <p className={styles.top_bar}>
      <span>{brand.brand}</span>
      {' | '}
      <span>{CATEGORY_LABELS_BY_CATEGORY[metadata.displayCategory]}</span>
    </p>
  )
}

function Title(props: LabwareCardProps) {
  const { loadName } = props.definition.parameters
  const { displayName } = props.definition.metadata

  return (
    <Link to={`${getPublicPath()}${loadName}`}>
      <h2 className={styles.title}>
        {displayName}
        <Icon className={styles.title_icon} name="chevron-right" />
      </h2>
    </Link>
  )
}

function Dimensions(props: LabwareCardProps) {
  const { definition } = props
  const { displayCategory } = definition.metadata
  const { xDimension, yDimension, zDimension } = definition.dimensions
  const dimsLabel =
    LABWARE_DIMS_BY_CATEGORY[displayCategory] || LABWARE_DIMS_BY_CATEGORY.other

  const dimensions = [
    { label: SHORT_X_DIM, value: toFixed(xDimension) },
    { label: SHORT_Y_DIM, value: toFixed(yDimension) },
    { label: SHORT_Z_DIM, value: toFixed(zDimension) },
  ]

  return (
    <LabeledValueTable
      className={styles.dimensions}
      direction={TABLE_ROW}
      label={
        <>
          {dimsLabel} <span className={styles.units}>({MM})</span>
        </>
      }
      values={dimensions}
    />
  )
}

function Wells(props: LabwareCardProps) {
  const { wells, metadata } = props.definition
  const { displayCategory } = metadata

  return (
    <div className={styles.wells}>
      <LabelText position={LABEL_TOP}>
        {`${NUM_WELLS_SHORT_BY_CATEGORY[displayCategory]}`}
      </LabelText>
      <Value>{Object.keys(wells).length}</Value>
    </div>
  )
}

function WellProperties(props: LabwareCardProps) {
  const { definition } = props
  const { displayCategory, displayVolumeUnits } = definition.metadata
  const wellProps = getUniqueWellProperties(definition)
  const wellDimsLabel =
    WELL_DIMS_BY_CATEGORY[displayCategory] || WELL_DIMS_BY_CATEGORY.other

  return (
    <div className={styles.wells}>
      {wellProps.map((w, i) => {
        const vol = getDisplayVolume(w.totalLiquidVolume, displayVolumeUnits, 2)
        const dims = [
          { label: DEPTH, value: toFixed(w.depth) },
          w.diameter != null
            ? { label: DIAMETER, value: toFixed(w.diameter) }
            : null,
          w.xDimension != null
            ? { label: X_DIM, value: toFixed(w.xDimension) }
            : null,
          w.yDimension != null
            ? { label: Y_DIM, value: toFixed(w.yDimension) }
            : null,
        ].filter(Boolean)

        return (
          <div key={i} className={styles.well_group}>
            <LabeledValueTable
              className={styles.well_dimensions}
              label={
                <>
                  {wellDimsLabel} <span className={styles.units}>({MM})</span>
                </>
              }
              values={dims}
            />
            <div className={styles.well_volume}>
              <LabelText position={LABEL_TOP}>{MAX_VOLUME}</LabelText>
              <Value>
                {vol} {displayVolumeUnits}
              </Value>
            </div>
          </div>
        )
      })}
    </div>
  )
}
