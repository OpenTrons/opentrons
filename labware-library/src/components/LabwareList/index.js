// @flow
// main LabwareList component
import * as React from 'react'

import styles from './styles.css'

import { getAllDefinitions } from '../../definitions'
import { FILTER_OFF } from '../../filters'
import LabwareCard from './LabwareCard'
import NoResults from './NoResults'

import type { FilterParams } from '../../types'

const filterMatches = (filter: ?string, value: string): boolean =>
  !filter || filter === FILTER_OFF || filter === value

export type LabwareListProps = {
  filters: FilterParams,
}

export default function LabwareList(props: LabwareListProps) {
  const { category, manufacturer } = props.filters
  const definitions = getAllDefinitions().filter(
    d =>
      filterMatches(category, d.metadata.displayCategory) &&
      filterMatches(manufacturer, d.brand.brand)
  )

  return definitions.length === 0 ? (
    <NoResults />
  ) : (
    <ul className={styles.list}>
      {definitions.map(d => (
        <LabwareCard key={d.otId} definition={d} />
      ))}
    </ul>
  )
}

export { NoResults }
