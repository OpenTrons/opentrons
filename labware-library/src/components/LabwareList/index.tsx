// main LabwareList component
import { getFilteredDefinitions } from '../../filters'
import type { FilterParams } from '../../types'
import { CustomLabwareCard } from './CustomLabwareCard'
import { LabwareCard } from './LabwareCard'
import styles from './styles.css'
import { getLabwareDefURI } from '@opentrons/shared-data'
import * as React from 'react'

export interface LabwareListProps {
  filters: FilterParams
}

export function LabwareList(props: LabwareListProps): JSX.Element {
  const definitions = getFilteredDefinitions(props.filters)

  return (
    <ul className={styles.list}>
      {definitions.map(d => (
        <LabwareCard key={getLabwareDefURI(d)} definition={d} />
      ))}

      <CustomLabwareCard isResultsEmpty={definitions.length === 0} />
    </ul>
  )
}
