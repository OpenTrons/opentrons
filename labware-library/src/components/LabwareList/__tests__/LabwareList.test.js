// @flow
// LabwareList tests
import * as React from 'react'
import Renderer from 'react-test-renderer'

import LabwareList from '../LabwareList'
import {getAllDefinitions} from '../../../definitions'

jest.mock('../../../definitions', () => ({getAllDefinitions: jest.fn()}))

const mockLabware = [
  require('@opentrons/shared-data/definitions2/generic_96_wellplate_380_ul.json'),
  require('@opentrons/shared-data/definitions2/opentrons_96_tiprack_300_ul.json'),
  require('@opentrons/shared-data/definitions2/usa_scientific_12_trough_22_ml.json'),
]

describe('LabwareList', () => {
  beforeAll(() => {
    ;(getAllDefinitions: any).mockReturnValue(mockLabware)
  })

  beforeEach(() => {
    jest.clearAllMocks()
  })

  test('component renders', () => {
    const tree = Renderer.create(<LabwareList />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('renders a <ul>', () => {
    const tree = Renderer.create(<LabwareList />)

    const items = tree.root.findAllByType('ul')
    expect(items).toHaveLength(1)
  })

  test('renders a <li> per labware definition', () => {
    const tree = Renderer.create(<LabwareList />)

    const items = tree.root.findAllByType('li')
    expect(items).toHaveLength(mockLabware.length)
  })
})
