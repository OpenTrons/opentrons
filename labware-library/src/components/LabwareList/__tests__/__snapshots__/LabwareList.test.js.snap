// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LabwareList component renders 1`] = `
<ul
  className="list"
>
  <LabwareCard
    definition={
      Object {
        "brand": Object {
          "brand": "generic",
        },
        "cornerOffsetFromSlot": Object {
          "x": 0,
          "y": 0,
          "z": 0,
        },
        "deprecated": false,
        "dimensions": Object {
          "xDimension": 127.76,
          "yDimension": 85.48,
          "zDimension": 64.48,
        },
        "metadata": Object {
          "displayCategory": "tubeRack",
          "displayName": "Fake Irregular Container",
          "displayVolumeUnits": "mL",
          "tags": Array [
            "fake",
            "opentrons",
          ],
        },
        "ordering": Array [
          Array [
            "A1",
            "B1",
            "C1",
            "E1",
            "G1",
            "I1",
          ],
          Array [
            "A2",
            "B2",
            "C2",
            "E2",
            "G2",
            "I2",
          ],
          Array [
            "A3",
            "B3",
            "C3",
            "E3",
            "G3",
            "I3",
          ],
          Array [
            "A4",
            "B4",
            "C4",
            "E4",
            "G4",
            "I4",
          ],
          Array [
            "A5",
            "B5",
            "C5",
            "E5",
            "G5",
            "I5",
          ],
          Array [
            "A6",
            "C6",
            "E6",
            "G6",
            "I6",
          ],
          Array [
            "A7",
            "C7",
            "E7",
            "G7",
            "I7",
          ],
          Array [
            "A8",
            "C8",
            "E8",
            "G8",
            "I8",
          ],
          Array [
            "A9",
            "C9",
            "E9",
            "G9",
            "I9",
          ],
          Array [
            "A10",
            "C10",
            "E10",
            "G10",
            "I10",
          ],
        ],
        "otId": "mock-id",
        "parameters": Object {
          "format": "irregular",
          "isMagneticModuleCompatible": false,
          "isTiprack": false,
          "loadName": "generic_50x3_ml_5x10_ml_tuberack",
        },
        "wells": Object {
          "A1": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 10,
            "y": 30,
            "z": 58.94,
          },
          "A10": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 100,
            "y": 30,
            "z": 58.94,
          },
          "A2": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 20,
            "y": 30,
            "z": 58.94,
          },
          "A3": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 30,
            "y": 30,
            "z": 58.94,
          },
          "A4": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 40,
            "y": 30,
            "z": 58.94,
          },
          "A5": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 50,
            "y": 30,
            "z": 58.94,
          },
          "A6": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 60,
            "y": 30,
            "z": 58.94,
          },
          "A7": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 70,
            "y": 30,
            "z": 58.94,
          },
          "A8": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 80,
            "y": 30,
            "z": 58.94,
          },
          "A9": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 90,
            "y": 30,
            "z": 58.94,
          },
          "B1": Object {
            "depth": 20.54,
            "diameter": 15,
            "shape": "circular",
            "totalLiquidVolume": 10000,
            "x": 15,
            "y": 15,
            "z": 48.94,
          },
          "B2": Object {
            "depth": 20.54,
            "diameter": 15,
            "shape": "circular",
            "totalLiquidVolume": 10000,
            "x": 25,
            "y": 15,
            "z": 48.94,
          },
          "B3": Object {
            "depth": 20.54,
            "diameter": 15,
            "shape": "circular",
            "totalLiquidVolume": 10000,
            "x": 35,
            "y": 15,
            "z": 48.94,
          },
          "B4": Object {
            "depth": 20.54,
            "diameter": 15,
            "shape": "circular",
            "totalLiquidVolume": 10000,
            "x": 45,
            "y": 15,
            "z": 48.94,
          },
          "B5": Object {
            "depth": 20.54,
            "diameter": 15,
            "shape": "circular",
            "totalLiquidVolume": 10000,
            "x": 55,
            "y": 15,
            "z": 48.94,
          },
          "C1": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 10,
            "y": 25,
            "z": 58.94,
          },
          "C10": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 100,
            "y": 25,
            "z": 58.94,
          },
          "C2": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 20,
            "y": 25,
            "z": 58.94,
          },
          "C3": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 30,
            "y": 25,
            "z": 58.94,
          },
          "C4": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 40,
            "y": 25,
            "z": 58.94,
          },
          "C5": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 50,
            "y": 25,
            "z": 58.94,
          },
          "C6": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 60,
            "y": 25,
            "z": 58.94,
          },
          "C7": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 70,
            "y": 25,
            "z": 58.94,
          },
          "C8": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 80,
            "y": 25,
            "z": 58.94,
          },
          "C9": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 90,
            "y": 25,
            "z": 58.94,
          },
          "E1": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 10,
            "y": 20,
            "z": 58.94,
          },
          "E10": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 100,
            "y": 20,
            "z": 58.94,
          },
          "E2": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 20,
            "y": 20,
            "z": 58.94,
          },
          "E3": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 30,
            "y": 20,
            "z": 58.94,
          },
          "E4": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 40,
            "y": 20,
            "z": 58.94,
          },
          "E5": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 50,
            "y": 20,
            "z": 58.94,
          },
          "E6": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 60,
            "y": 20,
            "z": 58.94,
          },
          "E7": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 70,
            "y": 20,
            "z": 58.94,
          },
          "E8": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 80,
            "y": 20,
            "z": 58.94,
          },
          "E9": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 90,
            "y": 20,
            "z": 58.94,
          },
          "G1": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 10,
            "y": 15,
            "z": 58.94,
          },
          "G10": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 100,
            "y": 15,
            "z": 58.94,
          },
          "G2": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 20,
            "y": 15,
            "z": 58.94,
          },
          "G3": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 30,
            "y": 15,
            "z": 58.94,
          },
          "G4": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 40,
            "y": 15,
            "z": 58.94,
          },
          "G5": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 50,
            "y": 15,
            "z": 58.94,
          },
          "G6": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 60,
            "y": 15,
            "z": 58.94,
          },
          "G7": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 70,
            "y": 15,
            "z": 58.94,
          },
          "G8": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 80,
            "y": 15,
            "z": 58.94,
          },
          "G9": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 90,
            "y": 15,
            "z": 58.94,
          },
          "I1": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 10,
            "y": 10,
            "z": 58.94,
          },
          "I10": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 100,
            "y": 10,
            "z": 58.94,
          },
          "I2": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 20,
            "y": 10,
            "z": 58.94,
          },
          "I3": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 30,
            "y": 10,
            "z": 58.94,
          },
          "I4": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 40,
            "y": 10,
            "z": 58.94,
          },
          "I5": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 50,
            "y": 10,
            "z": 58.94,
          },
          "I6": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 60,
            "y": 10,
            "z": 58.94,
          },
          "I7": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 70,
            "y": 10,
            "z": 58.94,
          },
          "I8": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 80,
            "y": 10,
            "z": 58.94,
          },
          "I9": Object {
            "depth": 10.54,
            "diameter": 10,
            "shape": "circular",
            "totalLiquidVolume": 3000,
            "x": 90,
            "y": 10,
            "z": 58.94,
          },
        },
      }
    }
    key="mock-id"
  />
  <LabwareCard
    definition={
      Object {
        "brand": Object {
          "brand": "opentrons",
          "brandId": Array [
            "t40u9sernisofsea",
          ],
        },
        "cornerOffsetFromSlot": Object {
          "x": 0,
          "y": 0,
          "z": 0,
        },
        "deprecated": false,
        "dimensions": Object {
          "xDimension": 127.76,
          "yDimension": 85.48,
          "zDimension": 50,
        },
        "metadata": Object {
          "displayCategory": "wellPlate",
          "displayName": "fake labware",
          "displayVolumeUnits": "µL",
        },
        "ordering": Array [
          Array [
            "A1",
          ],
          Array [
            "A2",
          ],
        ],
        "otId": "mock-id",
        "parameters": Object {
          "format": "96Standard",
          "isMagneticModuleCompatible": false,
          "isTiprack": false,
          "loadName": "opentrons_2_wellplate_100_ul",
        },
        "wells": Object {
          "A1": Object {
            "depth": 40,
            "diameter": 30,
            "shape": "circular",
            "totalLiquidVolume": 100,
            "x": 10,
            "y": 10,
            "z": 15,
          },
          "A2": Object {
            "depth": 40,
            "diameter": 30,
            "shape": "circular",
            "totalLiquidVolume": 100,
            "x": 20,
            "y": 10,
            "z": 15,
          },
        },
      }
    }
    key="mock-id"
  />
  <LabwareCard
    definition={
      Object {
        "brand": Object {
          "brand": "generic",
        },
        "cornerOffsetFromSlot": Object {
          "x": 77.76,
          "y": 35.48,
          "z": 0,
        },
        "deprecated": false,
        "dimensions": Object {
          "xDimension": 50,
          "yDimension": 50,
          "zDimension": 50,
        },
        "metadata": Object {
          "displayCategory": "wellPlate",
          "displayName": "fake labware 2",
          "displayVolumeUnits": "mL",
        },
        "ordering": Array [
          Array [
            "A1",
            "B1",
            "C1",
          ],
          Array [
            "A2",
            "B2",
            "C2",
          ],
        ],
        "otId": "mock-id",
        "parameters": Object {
          "format": "irregular",
          "isMagneticModuleCompatible": false,
          "isTiprack": false,
          "loadName": "generic_6_wellplate_1_ml",
        },
        "wells": Object {
          "A1": Object {
            "depth": 40,
            "diameter": 30,
            "shape": "circular",
            "totalLiquidVolume": 1000,
            "x": 10,
            "y": 28,
            "z": 10,
          },
          "A2": Object {
            "depth": 40,
            "diameter": 30,
            "shape": "circular",
            "totalLiquidVolume": 1000,
            "x": 19,
            "y": 28,
            "z": 10,
          },
          "B1": Object {
            "depth": 40,
            "diameter": 30,
            "shape": "circular",
            "totalLiquidVolume": 1000,
            "x": 10,
            "y": 19,
            "z": 10,
          },
          "B2": Object {
            "depth": 40,
            "diameter": 30,
            "shape": "circular",
            "totalLiquidVolume": 1000,
            "x": 19,
            "y": 19,
            "z": 10,
          },
          "C1": Object {
            "depth": 40,
            "diameter": 30,
            "shape": "circular",
            "totalLiquidVolume": 1000,
            "x": 10,
            "y": 10,
            "z": 10,
          },
          "C2": Object {
            "depth": 40,
            "diameter": 30,
            "shape": "circular",
            "totalLiquidVolume": 1000,
            "x": 19,
            "y": 10,
            "z": 10,
          },
        },
      }
    }
    key="mock-id"
  />
</ul>
`;
